File: java/eyes.playwright.java/src/main/java/com/applitools/eyes/playwright/universal/driver/SpecDriverPlaywright.java
package com.applitools.eyes.playwright.universal.driver;

import com.applitools.eyes.playwright.universal.Refer;
import com.applitools.eyes.playwright.universal.driver.dto.DriverInfoDto;
import com.applitools.eyes.playwright.universal.dto.Context;
import com.applitools.eyes.playwright.universal.dto.Element;
import com.applitools.eyes.playwright.universal.dto.Selector;
import com.applitools.eyes.universal.ISpecDriver;
import com.applitools.eyes.universal.Reference;
import com.applitools.eyes.universal.driver.ICookie;
import com.applitools.eyes.universal.dto.*;
import com.microsoft.playwright.*;
import com.microsoft.playwright.impl.PageImpl;
import com.microsoft.playwright.options.Cookie;

import java.lang.reflect.Method;
import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class SpecDriverPlaywright implements ISpecDriver {

    private Refer refer;

    public SpecDriverPlaywright(Refer ref) {
        this.refer = ref;
    }

    @Override
    public Boolean isDriver(Object driver) {
        return driver instanceof Page;
    }

    @Override
    public Boolean isContext(Reference context) {
        return ISpecDriver.super.isContext(context);
    }

    @Override
    public Boolean isElement(Reference element) {
        return ISpecDriver.super.isElement(element);
    }

    @Override
    public Boolean isSelector(Reference selector) {
        return ISpecDriver.super.isSelector(selector);
    }

    @Override
    public Context mainContext(Reference context) {
        Frame mainFrame = extractContextUtil(context);
        while (mainFrame.parentFrame() != null) {
            mainFrame = mainFrame.parentFrame();
        }

        Context frameContext = new Context();
        frameContext.setApplitoolsRefId(refer.ref(mainFrame, context));
        return frameContext;
    }

    @Override
    public Context parentContext(Reference context) {
        Frame frame = extractContextUtil(context);

        if (frame.parentFrame() == null) {
            return null;
        }

        Context frameContext = new Context();
        frameContext.setApplitoolsRefId(refer.ref(frame.parentFrame(), context));
        return frameContext;
    }

    @Override
    public Context childContext(Reference context, Reference element) {
        Object root = refer.deref(element);

        if (root instanceof ElementHandle) {
            Context frameContext = new Context();
            Frame frame = ((ElementHandle) root).contentFrame();
            frameContext.setApplitoolsRefId(refer.ref(frame, element));

            return frameContext;
        }

        return null;
    }

    @Override
    public Object executeScript(Reference context, String script, Object arg) {
        Object ctx = refer.deref(context);
        JSHandle res = null;

        Object args = derefArgsUtil(arg);

        if (ctx instanceof Frame) {
            res = ((Frame) ctx).evaluateHandle(script, args);
        } else if (ctx instanceof Page) {
            res = ((Page) ctx).evaluateHandle(script, args);
        }

        return handlerToObjectUtil(res, context);
    }

    @Override
    public Element findElement(Reference driver, Reference selector, Reference parent) {
        Object context = refer.deref(driver);
        Object root = parent == null? context : refer.deref(parent);
        Object locator = refer.deref(selector);

        ElementHandle elementHandle = null;
        String _selector = ((Selector)selector).getSelector();
        if (locator instanceof Locator) {
            elementHandle = ((Locator) locator).elementHandle();
        } else if (root instanceof Frame) {
            elementHandle = ((Frame) root).querySelector(_selector);
        } else if (root instanceof Page) {
            elementHandle = ((Page) root).querySelector(_selector);
        }

        if (elementHandle == null) {
            return null;
        }

        Element element = new Element();
        element.setApplitoolsRefId(refer.ref(elementHandle, driver));
        return element;
    }

    @Override
    public List<Reference> findElements(Reference context, Reference selector, Reference parent) {
        Object ctx = refer.deref(context);
        Object root = parent == null? ctx : refer.deref(parent);
        Object locator = refer.deref(selector);

        List<ElementHandle> elementHandles = new ArrayList<>();
        String _selector = ((Selector)selector).getSelector();
        if (locator instanceof Locator) {
            elementHandles = ((Locator) locator).elementHandles();
        } else if (root instanceof Frame) {
            elementHandles = ((Frame) root).querySelectorAll(_selector);
        } else if (root instanceof Page) {
            elementHandles = ((Page) root).querySelectorAll(_selector);
        }

        return elementHandles.stream()
                .map(elementHandle -> {
                    Element element = new Element(elementHandle);
                    element.setApplitoolsRefId(refer.ref(elementHandle, context));
                    return element;
                })
                .collect(Collectors.toList());
    }

    @Override
    public void setViewportSize(Reference driver, RectangleSizeDto windowSize) {
        Object page = refer.deref(driver);
        ((Page) page).setViewportSize(windowSize.getWidth(), windowSize.getHeight());
    }

    @Override
    public RectangleSizeDto getViewportSize(Reference driver) {
        Object page = refer.deref(driver);
        RectangleSizeDto viewportSize = new RectangleSizeDto();
        viewportSize.setWidth(((Page) page).viewportSize().width);
        viewportSize.setHeight(((Page) page).viewportSize().height);
        return viewportSize;
    }

    @Override
    public List<ICookie> getCookies(Reference driver, Reference context) {
        Page ctx = (Page) refer.deref(driver);
        List<Cookie> cookies = ctx.context().cookies();

        return cookies.stream()
                .map(TCookie::new)
                .collect(Collectors.toList());
    }

    @Override
    public DriverInfoDto getDriverInfo(Reference driver) {
        return new DriverInfoDto();
    }

    @Override
    public String getTitle(Reference driver) {
        Page context = (Page) refer.deref(driver);
        return context.title();
    }

    @Override
    public String getUrl(Reference driver) {
        Page context = (Page) refer.deref(driver);
        return context.url();
    }

    @Override
    public void visit(Reference driver, String url) {
        Page context = (Page) refer.deref(driver);
        context.navigate(url);
    }

    @Override
    public byte[] takeScreenshot(Reference driver) {
        Page context = (Page) refer.deref(driver);
        return context.screenshot();
    }

    static public String[] getMethodNames() {
        Method[] methods = SpecDriverPlaywright.class.getDeclaredMethods();
        return Stream.of(methods)
                .map(Method::getName)
                .filter(name -> !name.contains("getMethodNames") && !name.contains("Util"))
                .toArray(String[]::new);
    }

    private Object handlerToObjectUtil(JSHandle jsHandle, Reference context) {
        if (jsHandle == null) {
            return null;
        }

        String type = "";
        Pattern pattern = Pattern.compile("(?:.+@)?(\\w*)(?:\\(\\d+\\))?", Pattern.CASE_INSENSITIVE);
        Matcher matcher = pattern.matcher(jsHandle.toString());
        if (matcher.find()) {
            type = matcher.group(0).toLowerCase();
        }

        if (type.matches("array\\(\\d+\\)")) {
            Map<String, JSHandle> map = jsHandle.getProperties();
            List<Object> arrayValues = new ArrayList<>();
            for (JSHandle jsHandle1: map.values()) {
                arrayValues.add(handlerToObjectUtil(jsHandle1, context));
            }
            return arrayValues;
        } else if (type.equals("object")) {
            Map<String, JSHandle> map = jsHandle.getProperties();
            Map<String, Object> resultMap = new HashMap<>();
            for (Map.Entry<String, JSHandle> entry: map.entrySet()) {
                resultMap.put(entry.getKey(), handlerToObjectUtil(entry.getValue(), context));
            }
            return resultMap;
        } else if (type.equals("jshandle@node")) {
            Element element = new Element();
            element.setApplitoolsRefId(refer.ref(jsHandle.asElement(), context));
            return element;
        }

        return jsHandle.jsonValue();
    }

    private Object derefArgsUtil(Object arg) {
        if (arg == null) {
            return null;
        }
        ArrayList<Object> derefArg = new ArrayList<>();

        if (arg instanceof ArrayList) {
            for (Object argument: (ArrayList) arg) {
                if (argument instanceof ArrayList) {
                    derefArg.add(derefArgsUtil(argument));
                } else {
                    derefArg.add(refer.deref(argument));
                }
            }
            return derefArg;
        } else if (arg instanceof HashMap){
            HashMap<Object, Object> map = new HashMap<>();
            for (Map.Entry<?, ?> entry : ((HashMap<?, ?>) arg).entrySet()) {
                map.put(entry.getKey(), refer.deref(entry.getValue()));
            }
            return map;
        }

        return refer.deref(arg);
    }

    private Frame extractContextUtil(Reference context) {
        Object root = refer.deref(context);
        return isDriver(root) ? ((PageImpl) root).mainFrame() : (Frame) root;
    }
}

File: dotnet/Eyes.Playwright.DotNet/Universal/Driver/SpecDriverPlaywright.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;
using Applitools.Playwright.Universal.Driver.Dto;
using Applitools.Playwright.Universal.Dto;
using Applitools.Universal;
using Applitools.Universal.Driver;
using Applitools.Utils.Geometry;
using Microsoft.Playwright;

namespace Applitools.Playwright.Universal.Driver
{
    public class SpecDriverPlaywright : ISpecDriver
    {
        private readonly Refer refer_;

        public SpecDriverPlaywright(Refer refer)
        {
            refer_ = refer;
        }

        public bool IsDriver(object driver)
        {
            return driver is IPage;
        }

        public bool IsContext(Reference context)
        {
            throw new NotSupportedException("IsContext is not supported");
        }

        public bool IsElement(Reference element)
        {
            throw new NotSupportedException("IsElement is not supported");
        }

        public bool IsSelector(Reference selector)
        {
            throw new NotSupportedException("IsSelector is not supported");
        }

        public bool IsEqualElements()
        {
            throw new NotSupportedException("IsEqualElements is not supported");
        }

        public Reference MainContext(Reference context)
        {
            IFrame mainFrame = ExtractContextUtil_(context);
            while (mainFrame.ParentFrame != null)
            {
                mainFrame = mainFrame.ParentFrame;
            }

            Context frameContext = new Context
            {
                ApplitoolsRefId = refer_.Ref(mainFrame, context)
            };
            return frameContext;
        }

        public Reference ParentContext(Reference context)
        {
            IFrame frame = ExtractContextUtil_(context);

            if (frame.ParentFrame == null)
            {
                return null;
            }

            Context frameContext = new Context
            {
                ApplitoolsRefId = refer_.Ref(frame.ParentFrame, context)
            };
            return frameContext;
        }

        public Reference ChildContext(Reference context, Reference element)
        {
            object root = refer_.Deref(element);

            if (root is IElementHandle eh)
            {
                Context frameContext = new Context();
                IFrame frame = eh.ContentFrameAsync().GetAwaiter().GetResult();
                frameContext.ApplitoolsRefId = refer_.Ref(frame, element);

                return frameContext;
            }

            return null;
        }

        public object ExecuteScript(Reference context, string script, object arg)
        {
            object ctx = refer_.Deref(context);
            IJSHandle res = null;

            object args = DerefArgsUtil(arg);

            if (ctx is IFrame frame)
            {
                res = frame.EvaluateHandleAsync(script, args).GetAwaiter().GetResult();
            }
            else if (ctx is IPage page)
            {
                res = page.EvaluateHandleAsync(script, args).GetAwaiter().GetResult();
            }

            return HandlerToObjectUtil(res, context);
        }

        public Reference FindElement(Reference driver, Reference selector, Reference parent)
        {
            object context = refer_.Deref(driver);
            object root = parent == null ? context : refer_.Deref(parent);
            object locator = refer_.Deref(selector);

            IElementHandle elementHandle = null;
            string selectorPath = ((Selector)selector).SelectorPath;
            if (locator is ILocator loc)
            {
                elementHandle = loc.ElementHandleAsync().GetAwaiter().GetResult();
            }
            else if (root is IFrame frame)
            {
                elementHandle = frame.QuerySelectorAsync(selectorPath).GetAwaiter().GetResult();
            }
            else if (root is IPage page)
            {
                elementHandle = page.QuerySelectorAsync(selectorPath).GetAwaiter().GetResult();
            }

            if (elementHandle == null)
            {
                return null;
            }

            Element element = new Element
            {
                ApplitoolsRefId = refer_.Ref(elementHandle, driver)
            };
            return element;
        }

        public List<Reference> FindElements(Reference driver, Reference selector, Reference parent)
        {
            object context = refer_.Deref(driver);
            object root = parent == null ? context : refer_.Deref(parent);
            object locator = refer_.Deref(selector);

            IEnumerable<IElementHandle> elementHandles = null;
            string selectorPath = ((Selector)selector).SelectorPath;
            if (locator is ILocator loc)
            {
                elementHandles = loc.ElementHandlesAsync().GetAwaiter().GetResult();
            }
            else if (root is IFrame frame)
            {
                elementHandles = frame.QuerySelectorAllAsync(selectorPath).GetAwaiter().GetResult();
            }
            else if (root is IPage page)
            {
                elementHandles = page.QuerySelectorAllAsync(selectorPath).GetAwaiter().GetResult();
            }

            if (elementHandles == null)
            {
                return new List<Reference>();
            }

            List<Reference> result = elementHandles.Select<IElementHandle, Reference>(
                    eh =>
                    {
                        Element e = new Element(eh)
                        {
                            ApplitoolsRefId = refer_.Ref(eh, driver)
                        };
                        return e;
                    })
                .ToList();
            return result;
        }

        public void SetElementText()
        {
            throw new NotSupportedException("SetElementText is not supported");
        }

        public string GetElementText()
        {
            throw new NotSupportedException("GetElementText is not supported");
        }

        public void SetWindowSize(Reference driver, RectangleSize windowSize)
        {
            throw new NotSupportedException("SetWindowSize is not supported");
        }

        public RectangleSize GetWindowSize(Reference driver)
        {
            throw new NotSupportedException("GetWindowSize is not supported");
        }

        public void SetViewportSize(Reference driver, RectangleSize windowSize)
        {
            object page = refer_.Deref(driver);
            ((IPage)page).SetViewportSizeAsync(windowSize.Width, windowSize.Height);
        }

        public RectangleSize GetViewportSize(Reference driver)
        {
            object page = refer_.Deref(driver);
            RectangleSize viewportSize = new RectangleSize();
            PageViewportSizeResult pageVpSize = ((IPage)page).ViewportSize;
            viewportSize.Width = pageVpSize?.Width ?? 0;
            viewportSize.Height = pageVpSize?.Height ?? 0;
            return viewportSize;
        }

        public List<ICookie> GetCookies(Reference driver, Reference context)
        {
            IPage ctx = (IPage)refer_.Deref(driver);
            IReadOnlyList<BrowserContextCookiesResult> cookies = ctx.Context.CookiesAsync().GetAwaiter().GetResult();

            return cookies.Select<BrowserContextCookiesResult, ICookie>(cookie => new TCookie(cookie)).ToList();
        }

        public object GetDriverInfo(Reference driver)
        {
            return new DriverInfoDto();
        }

        public Dictionary<string, object> GetCapabilities(Reference driver)
        {
            throw new NotSupportedException("GetCapabilities is not supported");
        }

        public string GetTitle(Reference driver)
        {
            IPage context = (IPage)refer_.Deref(driver);
            return context.TitleAsync().GetAwaiter().GetResult();
        }

        public string GetUrl(Reference driver)
        {
            IPage context = (IPage)refer_.Deref(driver);
            return context.Url;
        }

        public void Click()
        {
            throw new NotSupportedException("Click is not supported");
        }

        public void Visit(Reference driver, string url)
        {
            IPage context = (IPage)refer_.Deref(driver);
            context.GotoAsync(url);
        }

        public void Hover(Reference context, Reference element)
        {
            throw new NotSupportedException("Hover is not supported");
        }

        public void ScrollIntoView(Reference context, Reference element, bool align)
        {
            throw new NotSupportedException("ScrollIntoView is not supported");
        }

        public void WaitUntilDisplayed(Reference context, Reference selector)
        {
            throw new NotSupportedException("WaitUntilDisplayed is not supported");
        }

        public byte[] TakeScreenshot(Reference driver)
        {
            IPage context = (IPage)refer_.Deref(driver);
            return context.ScreenshotAsync().GetAwaiter().GetResult();
        }
        
        private static readonly Regex jsHandlePattern_ = new Regex("(?:.+@)?(\\w*)(?:\\(\\d+\\))?",
            RegexOptions.Compiled | RegexOptions.IgnoreCase);

        private static readonly Regex arrayPattern_ = new Regex("array\\(\\d+\\)",
            RegexOptions.Compiled | RegexOptions.IgnoreCase);
        
        private object HandlerToObjectUtil(IJSHandle jsHandle, Reference context)
        {
            if (jsHandle == null)
            {
                return null;
            }

            if (jsHandle is IElementHandle)
            {
                Element element = new Element
                {
                    ApplitoolsRefId = refer_.Ref(jsHandle.AsElement(), context)
                };
                return element;
            }

            string type = "";

            Match jsHandleMatcher = jsHandlePattern_.Match(jsHandle.ToString());
            if (jsHandleMatcher.Success)
            {
                type = jsHandleMatcher.Groups[0].Value.ToLower();
            }

            if (arrayPattern_.Match(type).Success)
            {
                var map = jsHandle.GetPropertiesAsync().GetAwaiter().GetResult();
                var arrayValues = new List<object>();
                foreach (IJSHandle jsHandle1 in map.Values)
                {
                    arrayValues.Add(HandlerToObjectUtil(jsHandle1, context));
                }

                return arrayValues;
            }

            if (type.Equals("object"))
            {
                var map = jsHandle.GetPropertiesAsync().GetAwaiter().GetResult();
                var resultMap = new Dictionary<string, object>();
                foreach (var entry in map)
                {
                    resultMap.Add(entry.Key, HandlerToObjectUtil(entry.Value, context));
                }

                return resultMap;
            }

            return jsHandle.JsonValueAsync<object>().GetAwaiter().GetResult();
        }

        private object DerefArgsUtil(object arg)
        {
            if (arg == null)
            {
                return null;
            }

            var derefArg = new List<object>();

            if (arg is Dictionary<object, object> argsDict)
            {
                var map = new Dictionary<object, object>();
                foreach (var item in argsDict)
                {
                    map.Add(item.Key, DerefArgsUtil(item.Value));
                }

                return map;
            }

            if (arg is IList argsList)
            {
                foreach (object argument in argsList)
                {
                    if (argument is Reference)
                    {
                        derefArg.Add(refer_.Deref(argument));
                    }
                    else
                    {
                        derefArg.Add(DerefArgsUtil(argument));
                    }
                }

                return derefArg;
            }


            return refer_.Deref(arg);
        }

        private IFrame ExtractContextUtil_(Reference context)
        {
            object root = refer_.Deref(context);
            return IsDriver(root) ? ((IPage)root).MainFrame : (IFrame)root;
        }
    }
}